### ISSUE TITLE
Lossing High-Frenquency Details on Blended Region 

### ISSUE THREAD

**Issue by Gabriel-Huang** [matt-baugh/pytorch-poisson-image-editing#2]
First of all, thanks for the great work. I really appreciate it! 

I noticed that the image is losing high-frequency details after blending, any thoughts on this?

**Comment by matt-baugh** [comment]
What sort of blending method/settings are you using? If you're using mixed blending, it should maintain the stronger gradients which I would imagine keeps the high-frequency details.
Any image examples (including code if possible) would also help :)

**Comment by matt-baugh** [comment]
Sorry I was away for a while so slow to reply

**Comment by Gabriel-Huang** [comment]
Thank you very much for the reply!
I'm using the [blend](https://github.com/matt-baugh/pytorch-poisson-image-editing/blob/24826ecc5f4bb9ecf760919fcf01f27b5605a103/pietorch/functional.py#L14) func, and I turned off `mixed_blending`.

Here's an image example:
![target](https://github.com/matt-baugh/pytorch-poisson-image-editing/assets/22915903/e925331d-74cb-4a2e-b4f2-ce7591cc289b)
![source](https://github.com/matt-baugh/pytorch-poisson-image-editing/assets/22915903/a80975b6-7ff4-48a7-ac33-cfb87990dc11)
![mask](https://github.com/matt-baugh/pytorch-poisson-image-editing/assets/22915903/f7dbc89c-3bd4-4c29-8ecb-1cd4d8fb473c)
![result](https://github.com/matt-baugh/pytorch-poisson-image-editing/assets/22915903/15b73c04-56ab-4350-a384-7b432d91d108)

My intention is to blend this texture patch into the target image, the target and the texture are of the same size, so the corner_coord should be (0, 0).

The resulting image is slightly (barely noticeable) blurry than the original image, even for the region outside the mask. 

Any advice is appreciated!

-Gabriel

**Comment by matt-baugh** [comment]
Ah ok, when I switch back and forth between them I see what you mean.
I think that it may just be a limitation of how expressive the Fourier representation is, as that is where the blending is actually performed (a key part of the process is moving into fourier space, applying the Green's function, and then moving back) so it may be that some information is lost over that operation.

If you are using the source code locally, you could try increasing the size of the fourier signals used (s parameter here https://pytorch.org/docs/stable/generated/torch.fft.ifftn.html ) as that should give it a bit more capacity.

Alternatively, if you are ok temporarily converting your image to a numpy array, I found [blend_dst_numpy](https://github.com/matt-baugh/pytorch-poisson-image-editing/blob/master/pietorch/functional_numpy.py#L162) to work well even when applied to patches within the image (that one is the same algorithm OpenCV uses), so you could crop your image + mask to just the area you want to change, alter the corner_coord accordingly, and then apply that blend function.
That way you are guaranteed only to affect a bounding box around the area you are blending into, so if there is any blur it won't cover the entire image.
Hope this helps, happy to talk more if you have issues.

**Comment by Gabriel-Huang** [comment]
Thanks for the advice! 
I reintroduced the details into the blended image by incorporating the high-frequency residuals. The process involves:
1. Blending the target image with nothing, which yields a slightly blurred version.
2. Calculating the residuals between this blurred output and the original target image.

Below is a code snippet that demonstrates this, and can be added just before the final return statement:
```python
if restore_detail:
    laplacian_target = torch.sum(torch.stack([torch.gradient(grad, dim=grad_dim)[0]
                                for grad, grad_dim in zip(target_grads, chosen_dimensions)]), dim=0)
    init_blended_target = torch.fft.ifftn(torch.fft.fftn(laplacian_target, dim=chosen_dimensions) * green_function, dim=chosen_dimensions)

    integration_constant_target = init_blended_target[tuple([slice(1) if i in chosen_dimensions else slice(None)
                                            for i in range(num_dims)])]
    # Leave out padding + border, to avoid artefacts
    inner_blended_target = init_blended_target[tuple([slice(PAD_AMOUNT + 1, -PAD_AMOUNT - 1) if i in chosen_dimensions else slice(None)
                                    for i in range(num_dims)])]
    res_target = target.clone()
    res_target[res_indices] = torch.real(inner_blended_target - integration_constant_target)
    res_target = res_target.clip(0, 1)
    residuals = (target - res_target) * (1 - mask)

    result = (result + residuals).clip(0, 1)
```

**Comment by Clement-Lelievre** [comment]
hi @Gabriel-Huang 

I've tested your above snippet as I was running into the same blur issue, and in my case there is no difference pixelwise vs using the lib's `blend` method out of the box; is it expected that on some images your addition won't change anything to the blended image?

**Comment by LonglongaaaGo** [comment]
@Gabriel-Huang  Thank you so much for sharing. It works well. 
However, when I tested this on high-resolution images, it showed some exposure effects, and I had no idea what happened to that. Can you give some insights?
